common/src/main/java/org/conscrypt/OpenSSLAeadCipherAES.java

```diff
@@ -53,6 +53,60 @@ int getCipherBlockSize() {
         return AES_BLOCK_SIZE;
     }
 
+    @Override
+    protected AlgorithmParameterSpec getParameterSpec(AlgorithmParameters params)
+        throws InvalidAlgorithmParameterException {
+        if (params != null) {
+            AlgorithmParameterSpec spec = Platform.fromGCMParameters(params);
+            if (spec != null) {
+                return spec;
+            }
+            return super.getParameterSpec(params);
+        }
+        return null;
+    }
+
+    @Override
+    protected AlgorithmParameters engineGetParameters() {
+        // iv will be non-null after initialization.
+        if (iv == null) {
+            return null;
+        }
+
+        AlgorithmParameterSpec spec = Platform.toGCMParameterSpec(
+            tagLengthInBytes * 8, iv);
+        if (spec == null) {
+            // The platform doesn't support GCMParameterSpec. Fall back to
+            // the generic AES parameters so at least the caller can get the
+            // IV.
+            return super.engineGetParameters();
+        }
+
+        try {
+            AlgorithmParameters params = AlgorithmParameters.getInstance("GCM");
+            params.init(spec);
+            return params;
+        } catch (NoSuchAlgorithmException e) {
+            // We should not get here.
+            throw (Error) new AssertionError("GCM not supported").initCause(e);
+        } catch (InvalidParameterSpecException e) {
+            // This may happen since Conscrypt doesn't provide this itself.
+            return null;
+        }
+    }
+
+    @Override
+    int getOutputSizeForFinal(int inputLen) {
+        // For GCM, the tag is a fixed length and there is no padding or other
+        // concerns, so we can calculate the exact length required without a
+        // native call
+        if (isEncrypting()) {
+            return bufCount + inputLen + tagLengthInBytes;
+        } else {
+            return Math.max(0, bufCount + inputLen - tagLengthInBytes);
+        }
+    }
```

Comment:
Bonus points for a unit test for this

Reply:
There actually is one, it's just buried in CipherTest: https://github.com/google/conscrypt/blob/master/openjdk-integ-tests/src/test/java/org/conscrypt/javax/crypto/CipherTest.java#L1215